dev/DynamicDependency/DynamicDependencyLifetimeManagerShadow/winmain.cpp

```diff
@@ -0,0 +1,71 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+const HRESULT WINDOWSAPPSDK_DDLM_SHUTDOWN{ MAKE_HRESULT(SEVERITY_SUCCESS, FACILITY_ITF, 0x8000) };
+const HRESULT WINDOWSAPPSDK_DDLM_CALLER_TERMINATED{ MAKE_HRESULT(SEVERITY_SUCCESS, FACILITY_ITF, 0x8001) };
+const HRESULT WINDOWSAPPSDK_DDLM_CALLER_NOT_FOUND{ MAKE_HRESULT(SEVERITY_SUCCESS, FACILITY_ITF, 0x8002) };
+
+int WINAPI WinMain(HINSTANCE /*hInstance*/, HINSTANCE /*hPrevInstance*/, PSTR /*lpCmdLine*/, int /*nCmdShow*/)
+{
+    // Usage: This.Exe <eventname>
+    // where
+    //     eventname = event name signaling we should quit. Syntax: "<processid>;<packagefullname>;<uniqueid>"
+
+    // Parse the command line
+    const auto commandLine{ GetCommandLineW() };
+    RETURN_HR_IF_NULL(E_INVALIDARG, commandLine);
+    int argc{};
+    PWSTR* argv{ CommandLineToArgvW(commandLine, &argc) };
+    RETURN_HR_IF_NULL(E_INVALIDARG, argv);
+    wil::unique_hlocal argvBuffer{ argv };
+    RETURN_HR_IF(E_INVALIDARG, argc < 2);
+    PCWSTR eventName{ argv[1] };
+    PWSTR endPtr{};
+    uint32_t callerProcessId{ wcstoul(eventName, &endPtr, 10) };
+    // We'll never have ProcessId=0
+    RETURN_HR_IF_MSG(E_INVALIDARG, (callerProcessId == 0) || (endPtr == nullptr) || (*endPtr != L';'), "CommandLine=%ls", eventName);
+
+    // Open the named event created by the process calling the Bootstrap API (the process that caused us to run)
+    // If we can't it's because the process has terminated (or already called MddBootstrapShutdown())
+    // so we have nothing to do and we're out.
+    wil::unique_event_nothrow endOfTheLine;
+    RETURN_HR_IF_MSG(WINDOWSAPPSDK_DDLM_CALLER_NOT_FOUND, !endOfTheLine.try_open(eventName), "DDLM EventName=%ls", eventName);
+
+    // Open a handle to the calling process so we can watch for its termination
+    // (in case the process terminates without calling MddBootstrapShutdown())
+    wil::unique_handle callerProcessHandle{ OpenProcess(SYNCHRONIZE, FALSE, callerProcessId) };
+    RETURN_LAST_ERROR_IF_NULL(callerProcessHandle);
```

Comment:
consider standardized error reporting, via HRESULT_FROM_WIN32

Reply:
Error reporting _is_ standardized. OpenProcess is a BOOL/GetLastError Win32 API. `RETURN_LAST_ERROR_IF_NULL` returns `HRESULT_FROM_WIN32(GetLastError(x))` if x is null. Name might be a little misleading if you forget Win32 error land called it LastError but all good